pkg/process/process.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package process
+
+import (
+	"fmt"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/cgroup"
+)
+
+// key is a unique identifier for a process.
+// TODO(kakkoyun): This is not unique enough. We will need to add more fields.
+// For example, we can use the process start time to make sure we are not.
+// However, reading the stats file is expensive. For now, we will rely on the PID.
+// I will keep the struct to make it easier to add more fields in the future.
+type key struct {
+	pid int
+}
+
+type Tree struct {
+	tree map[key]*process
+}
+
+type process struct {
+	procfs.Proc
+
+	tree *Tree
+
+	parent int
+	// name of the field is the same as the one in kernel struct.
+	starttime uint64
+}
+
+func (p *process) ancestors() []*procfs.Proc {
+	ancestors := []*procfs.Proc{}
+	proc := p
+	for {
+		if proc.parent == 0 {
+			break
+		}
+		k := key{pid: proc.parent}
+		parent, ok := (*p.tree).tree[k]
+		if !ok {
+			break
+		}
+		ancestors = append(ancestors, &parent.Proc)
+		proc = parent
+	}
+	return ancestors
+}
+
+// NewTree returns a new process tree with current state of all the processes on the system.
+func NewTree() *Tree {
+	return &Tree{
+		// TODO(kakkoyun): This is an ever growing map. Introduce a mechanism to prune it.
+		tree: make(map[key]*process),
+	}
+}
+
+// FindAllAncestorProcessIDsInSameCgroup returns all ancestor process IDs for a given PID in the same cgroup.
+func (t *Tree) FindAllAncestorProcessIDsInSameCgroup(pid int) ([]int, error) {
```

Comment:
Are we 100% certain that there is no race conditions possible here? I feel like there could be multiple go routines potentially doing work here no?

Reply:
In theory, since we only have one profiler right now, there shouldn't be a race condition because we only have on `cpu.Run` at a time https://github.com/parca-dev/parca-agent/blob/68d2c8cf561d94ab765e1f93d9408938d0eae694/pkg/profiler/cpu/cpu.go#L395

However, with multiple profilers, this could be possible. But if it's the case, we have similar problems in several places in the code base.

I will test what you proposed above. With that, this could be ok.

Reply:
Ah also we have a lock on the call site https://github.com/parca-dev/parca-agent/blob/20ac55d6c10d6d518b79f24cca88c83655ea3e10/pkg/metadata/service_discovery.go#L35-L36

Reply:
see 432b2ff4139ee41590b3795e7b858f2c086e0767